src/Microsoft.Identity.Web/TokenAcquisition.cs

```diff
@@ -310,7 +337,23 @@ internal partial class TokenAcquisition : ITokenAcquisitionInternal
                 }
             }
 
-            return builder.ExecuteAsync();
+            try
+            {
+                return builder.ExecuteAsync();
+            }
+            catch (MsalServiceException exMsal) when (IsInvalidClientCertificateError(exMsal))
+            {
+                DefaultCertificateLoader.ResetCertificates(_microsoftIdentityOptions.ClientCertificates);
+                _application = null;
+
+                // Retry
+                retryClientCertificate = true;
+                return GetAuthenticationResultForAppAsync(scope, tenant, tokenAcquisitionOptions);
+            }
+            finally
+            {
+                retryClientCertificate = false;
```

Comment:
How often do we want to retry a failed cert? Is TokenAcquision a long lived service? If it is, when the cert expires, you reload it. The second time the cert expires, you will not reload it. Is this "reload once" the strategy you are aiming for?

Reply:
Although it's not the default, TokenAcquisition can be a singleton.
The reload once is to avoid an infinite loop. It's used in finally blocks with a tail recursion.